uknetdev, plat/kvm: Support for partial checksumming / checksum offloading

skuenzer opened the pull request:
For checksum offloading for TCP and UDP flows, typically, a partial checksum needs to be computed. This checksum covers only the pseudo header of the protocols. By specifying a start pointer and another pointer to the checksum field, a hardware network device is able to complete the computation of the checksum.
This PR introduces support for `lib/uknetdev` for such a feature. Two new flags and two fields for pbufs are introduced to specify information needed by the driver to offload the computation for a packet. Additionally, a new `uknetdev` driver state is introduced: `UK_NETDEV_UNPROBED`. This state indicates that network devices features weren't negotiated with the device yet. During device initialization, a call to `uk_netdev_probe()` is becoming mandatory to turn the device into `UK_NETDEV_UNCOFNIGURED` state. For some devices (like virtio-net) this is required to figure out if the backend supports partial checksumming.
This PR also implements partial checksumming for `virtio-net`.

skuenzer committed:
lib/uknetdev: Namespace netdev feature flags

This commit renames the netdev feature flags from a naming
scheme that was unspecific to uknetdev (`UK_FEATURE_*`) to
`UK_NETDEV_F_*`.
The intention is to avoid confusion and potential conflicts
with other Unikraft systems that introduce flags.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uknetdev: Support partial checksumming

In order to support partial checksummed packets for transmit and
receive, new netbuf flags are added:
* `UK_NETBUF_F_PARTIAL_CSUM`: Indicates that the the checksum field
  is only partially computed.
* `UK_NETBUF_F_DATA_VALID`: Indicates that the packet payload
  is checked and no data corruption happened although the checksum
  field might be incorrect due to checksum offloading.
Device drivers indicate the support for partial checksumming and the
support for these two flagson netbufs with the device feature flag
`UK_NETDEV_F_PARTIAL_CSUM`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
lib/uknetdev: Introduce feature probing

Introduces probing for netdev features. A new mandatory API
call `uk_netdev_probe()` is introduced that enables drivers
to negotiate network device features and options with the
backend or underlying device. Examples are maximum number
of queues or partial checksum offloading. During
initialization `uk_netdev_probe()` has to be called before
`uk_netdev_info_get()` or `uk_netdev_configure()` is called.
For drivers, the implementation of the probe function is
optional.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
drivers/virtio-net: Do not print hardware address

Do not print the hardware address during feature negotiation.
For the case that the hardware address is overwritten or is
not used by an actual Unikraft application or network stack,
this information might be misleading for users.
Suggestion is to print the address by the application or
network stack.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
drivers/virtio-net: Support partial checksumming

This commit implements partially checksummed netbufs for `virtio-net`
for sending and receiving. Support for partial checksumming is
negotiated during netdev probing. The reason is that some virtio
backends (e.g., qemu user netdev) do not support checksum offloading.
The feature availability has to be queried at run time. For this
purpose, `virtio_netdev_feature_negotiate()` is reorganized so that
it can be called via `uk_netdev_probe()`.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

craciunoiuc commented on lib/uknetdev/include/uk/netbuf.h:
```diff
@@ -101,10 +101,34 @@ typedef void (*uk_netbuf_dtor_t)(struct uk_netbuf *);
 	struct uk_netbuf *next;
 	struct uk_netbuf *prev;
 
+	uint8_t flags;         /**< Flags for this netbuf */
```
Is this flags element placed correctly? Considering 8 byte alignment (for 64bit), this might cause unnecessary padding.

If I remember correctly, the compiler might move elements in a structure to keep padding to a minimum. Is this true?

Furthermore, if padding is inevitable, maybe it's worth moving to `uint16_t` so the flags storage is more than enough in the future.

skuenzer replied:
Since this struct is never used to interact with an actual device directly, padding doesn't matter. A driver will parse the fields and will feed device specific descriptors. Whatever is making accessing fields in this struct faster is actually good for the networking subsystem - so I prefer leaving it up to the compiler. I don't expect many flags in the future but we can still extend its data type to `uint16_t` when needed since everything is recompiled together. The `uint8_t` was followed the `pbuf` example in lwip.

craciunoiuc replied:
Sounds good. I just wanted to highlight that we have space for more if needed 😉

craciunoiuc commented on lib/uknetdev/include/uk/netbuf.h:
```diff
@@ -520,6 +553,14 @@ static inline int uk_netbuf_header(struct uk_netbuf *head, int16_t len)
+
+	if (head->flags & UK_NETBUF_F_PARTIAL_CSUM) {
+		/* When header gets removed, delete PARTIAL_CSUM flag */
+		if (unlikely((len < 0) && (-len > head->csum_start)))
```
```suggestion
		if (unlikely((len < 0) && (len < head->csum_start)))
```
Isn't it more concise like this? That `-` might be easy to miss.

skuenzer replied:
This change is would not be the same expression anymore.
It is either
`-len > head->csum_start`
or
`len < -head->csum_start`
but in both cases one minus is staying.

craciunoiuc replied:
Whoops, you're right

craciunoiuc reviewed (commented):
Hey @skuenzer,

I looked over your commits and I left a couple of comments.
All in all, the patches look good. I will let @mogasergiu review the patches first on the `lwip` part.

There is a typo in the second commit description `flagson` -> `flags on`.

Afterwards I'll come back and we'll continue the review/approval.

Cezar

skuenzer committed:
drivers/virtio: Refactor feature bit macro

The virtio feature bit macros where misleading because one
could imply that multiple feature flags can be handed over
as logical OR chain. Because the macro just does one bit
shift operation, unexpected results occur in such
situation.
This commit renames the macro so that it is clear that only
one feature bit can be set, cleared or queried.

Checkpatch-Ignore: SINGLE_STATEMENT_DO_WHILE_MACRO
Checkpatch-Ignore: SPACING
Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

craciunoiuc commented on lib/uknetdev/include/uk/netbuf.h:
```diff
@@ -101,10 +101,34 @@ typedef void (*uk_netbuf_dtor_t)(struct uk_netbuf *);
+ * interface card when leaving the host.
+ * NOTE: The defacto standard is that IP header checksums are always
+ * computed in software). In the case of TCP and UDP, a softrware computed
+ * partial checksum (just the pdeudo header) is filled into the checksum
```
There is a typo here: `pdeudo` -> `pseudo`

craciunoiuc commented on plat/drivers/virtio/virtio_net.c:
```diff
@@ -372,8 +368,23 @@ static int virtio_netdev_xmit(struct uk_netdev *dev,
+	 * TODO: We assume that the PARTIAL_CSUM flag is only set with
+	 *       the first netbuf of a queue. If this is not the case,
+	 *       (e.g., due to encapsulation of protocol headers with
+	 *        prepending netbufs) we we need to replace the call
```
Duplicate `we` here.

craciunoiuc commented on plat/drivers/virtio/virtio_net.c:
```diff
@@ -830,10 +858,15 @@ static __u16 virtio_net_mtu_get(struct uk_netdev *n)
+static int virtio_netdev_feature_negotiate(struct uk_netdev *n)
 {
 	__u64 host_features = 0;
+	__u64 drv_features  = 0x0;
```
No need to write it in hex if it's 0.

skuenzer replied:
Sure not needed, I can remove it if you prefer - I don't mind. I find it supportive to the understanding that we are expecting bit flags to be set when using the `0x` notation - kind of: "this is not an ordinary number, we will do bit-wise operations on it."

craciunoiuc commented on plat/drivers/virtio/virtio_net.c:
```diff
@@ -843,17 +876,59 @@ static int virtio_netdev_feature_negotiate(struct virtio_net_device *vndev)
+
+	/**
+	 * Gratuitous ARP
+	 * NOTE: We tell that we will do gratuitous ARPs ourself.
```
Typo here: `ourself` -> `ourselves`

craciunoiuc reviewed (changes requested):
Hey @skuenzer,

I reviewed your new changes and came back with some more comments.

Still, the patch series looks good and it works for me. As soon as Sergiu finishes
his review I will also approve this.

Cezar

skuenzer commented:
The pushed update fixes the comments from your review, @craciunoiuc . Thanks for your feedback!

craciunoiuc reviewed (approved):
Hey @skuenzer @marcrittinghaus,

The PR looks good to me now. Please, @marcrittinghaus, take care when approving this PR such that it is introduced in the correct order with respect to the other PRs related (https://github.com/unikraft/lib-lwip/pull/12 https://github.com/unikraft/fork-lwip/pull/1 https://github.com/unikraft/lib-lwip/pull/10)

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@gmail.com>

skuenzer committed:
plat/xen: netfront: Query XenBus on device probe

Moves the device feature negotiation to device probing. This causes
XenBus interactions only when needed: Whenever a netfront device
is actually initialized for usage with the uknetdev API.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
plat/xen: netfront: Fix memory leak when parsing IP configuration

If parsing of a given XenStore IP configuration failed, the XenStore
response string `ip_str` needs still to be free'd.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

skuenzer committed:
plat/xen: netfront: Support partial checksumming

This commit enables handling of partially checksummed netbufs
for `netfront` for sending and receiving.

Signed-off-by: Simon Kuenzer <simon.kuenzer@neclab.eu>

marcrittinghaus reviewed (approved):
I did not do a detailed review but I have been using this code for quite some time internally and the code is of high quality. The PR depends on https://github.com/unikraft/lib-lwip/pull/10 being merged, too. Otherwise, network breaks.

Approved-by: Marc Rittinghaus <marc.rittinghaus@kit.edu>
